### ISSUE TITLE
When decode_times fails, warn rather than failing

### ISSUE THREAD

**Issue by efiring** [pydata/xarray#2848]
There are various reasons why time decoding can fail on a netcdf or opendap file, but the immediate solution is always to try again with `decode_times=False`.  Therefore it would save time and frustration if instead of failing, `open_dataset` would simply issue a warning for any variable it could not decode, and proceed to decode those that it can.  As an example, here is a URL for which the time coordinate can be converted but there is another time-related array, `tau` that cannot--and that probably shouldn't be, in any case:
```python
import xarray as xr
url = "http://tds.hycom.org/thredds/dodsC/GLBu0.08/expt_90.9/uv3z"
d0 = xr.open_dataset(url, decode_times=False)
print(d0.time)
print(d0.tau)
```
yielding
```
<xarray.DataArray 'time' (time: 438)>
array([105768., 105792., 105816., ..., 119472., 119496., 119520.])
Coordinates:
  * time     (time) float64 1.058e+05 1.058e+05 ... 1.195e+05 1.195e+05
Attributes:
    long_name:    Valid Time
    units:        hours since 2000-01-01 00:00:00
    time_origin:  2000-01-01 00:00:00
    calendar:     gregorian
    axis:         T
    NAVO_code:    13

<xarray.DataArray 'tau' (time: 438)>
array([0., 0., 0., ..., 0., 0., 0.])
Coordinates:
  * time     (time) float64 1.058e+05 1.058e+05 ... 1.195e+05 1.195e+05
Attributes:
    long_name:    Tau
    units:        hours since analysis
    time_origin:  2013-08-19 00:00:00
    NAVO_code:    56
```

#### Output of ``xr.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.7 | packaged by conda-forge | (default, Feb 28 2019, 02:16:08) 
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)]
python-bits: 64
OS: Darwin
OS-release: 16.7.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8
libhdf5: 1.10.2
libnetcdf: 4.6.1

xarray: 0.12.0
pandas: 0.24.2
numpy: 1.16.2
scipy: 1.2.1
netCDF4: 1.4.1
pydap: None
h5netcdf: 0.6.1
h5py: 2.8.0
Nio: None
zarr: None
cftime: 1.0.0
nc_time_axis: None
PseudonetCDF: None
rasterio: None
cfgrib: None
iris: None
bottleneck: 1.2.1
dask: 0.18.2
distributed: 1.22.0
matplotlib: 2.2.2
cartopy: 0.16.0
seaborn: None
setuptools: 40.0.0
pip: 18.1
conda: 4.5.11
pytest: 3.7.1
IPython: 6.5.0
sphinx: 1.8.0

</details>

**Comment by DocOtak** [comment]
When I was looking into this real quick after it was posted to the xarray mailing list, one of the things I attempted to do was use `xr.decode_cf()` on a DataArray object, which seems unsupported. I also found myself wanting some of the configuration options that the pandas [`read_csv()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) method has for decoding dates, particularly the ability to say which labels I want to decode and even method hooks for implementing my own parsing function if necessary. While in entirety would be way too complicated for xarray I think, it might be nice to emulate that API a little bit.

**Comment by fmaussion** [comment]
> it would save time and frustration if instead of failing, open_dataset would simply issue a warning for any variable it could not decode, and proceed to decode those that it can

I tend to agree, but changing this now would break all code which relies on `try ... except` logic to check if files can be decoded or not. I'm not sure it's worth changing it at this stage...

**Comment by efiring** [comment]
One possibility would be for the `decode_times` kwarg to accept a string 'warn' argument to select the new behavior.

**Comment by shoyer** [comment]
We could had add an overall argument for controlling this, e.g.,  `decoding_failure='error'` (default) vs  `decoding_failure='warn'` vs  `decoding_failure='silent'`. Though I'm not sure if these errors turn up practice with anything other than decoding times...

**Comment by max-sixty** [comment]
Trying to keep us below 1K issues — is this still current?

**Comment by efiring** [comment]
Yes, I don't see that anything has changed; but it would also be reasonable to say, "No one has submitted a PR, so this must not be high priority; go ahead and close it."

**Comment by ChrisBarker-NOAA** [comment]
No! please don't close this!

This just bit me again, and it's really annoying.

Example, from the FVCOM model:

```
	double time(time) ;
		time:long_name = "time" ;
		time:units = "seconds since 1858-11-17 00:00:00" ;
		time:format = "modified julian day (MJD)" ;
		time:time_zone = "UTC" ;
	int Itime(time) ;
		Itime:units = "days since 1858-11-17 00:00:00" ;
		Itime:format = "modified julian day (MJD)" ;
		Itime:time_zone = "UTC" ;
	int Itime2(time) ;
		Itime2:units = "msec since 00:00:00" ;
		Itime2:time_zone = "UTC" ;
```

what they have provided is a proper CF compliant time variable, but also two more -- `ITime` is the days since, and `ITime2` is milliseconds since the beginning of that day -- so you can combine to two to get milliseconds -- no idea why they did this, I can't imagine milliseconds matter to the model, but whatever. [NOTE: they used to use a float, rather than double for CF time, which didn't even give seconds precsion  :-( ]

Anyway -- no one wants to decode the Itime2 variable anyway -- but now that whole dataset barfs, and I have to either skip those variables, or not decode times, and then hand-decode the time variable.

WE could be even smarted, and respond differently if there is, e.g. a CF standard name of time, which should be decodable, or, for that matter, be more strict for the time coordinate, which also should be decodable.

In fact, I'm wondering how xarray decides that it *should* try to decode those variables? 

But that could get pretty complicated,  So I vote for:

> We could had add an overall argument for controlling this, e.g., decoding_failure='error' (default) vs decoding_failure='warn' vs decoding_failure='silent'.

or: 
> Though I'm not sure if these errors turn up practice with anything other than decoding times...

well, there's a decode_times parameter already, expecting True or False -- we could have it take:

`True  False, 'warn', or 'silent'`

As for: 
> "No one has submitted a PR, so this must not be high priority;

I'll do a PR if a core dev approves the approach.

NOTE: I'll need help with the type hints -- I hate type hints ....

**Comment by dcherian** [comment]
It's triggered by the "since" in `"msec since 00:00:00"` and (presumably) failing because `msec` is not UDUNITS-compliant. You could set decode_cf=False, edit the attr, then call `xr.decode_cf(ds)`

**Comment by kmuehlbauer** [comment]
The solution suggested by @dcherian above is the simplest for the informed end user. But fixing the `msec` -> `ms` is not sufficient, as `00:00:00` is no valid time string. This would have to be fixed too.

> well, there's a decode_times parameter already, expecting True or False -- we could have it take:

@ChrisBarker-NOAA That's not the whole story. Citing from [open_dataset](https://docs.xarray.dev/en/stable/generated/xarray.open_dataset.html):

> decode_times bool, CFDatetimeCoder or dict-like, optional
> 
> If True, decode times encoded in the standard NetCDF datetime format into datetime objects. Otherwise, use [coders.CFDatetimeCoder](https://docs.xarray.dev/en/stable/generated/xarray.coders.CFDatetimeCoder.html#xarray.coders.CFDatetimeCoder) or leave them encoded as numbers. Pass a mapping, e.g. {"my_variable": False}, to toggle this feature per-variable individually. This keyword may not be supported by all the backends.
>

So, users can deactivate any problematic variable from decoding, by passing a mapping as shown. Downstream devs could implement their own CFDatetimeCoder, with added functionality.

CFDatetimeCoder might actually be the right place to start, the suggested error-switch could just be added there (eg. `on_error` : `{"raise", "warn", "ignore"}` with `raise` as default). The underlying implementation is NOT straightforward as we already have several warnings issued on the way (numpy.datetime vs cftime, calendar issues etc).

For the user this might seem like a small step, but implementation wise it could be a giant leap.

**Comment by ChrisBarker-NOAA** [comment]
First: acknowledging that the source of the problem (and others like it) is that the data provider is providing bad data -- don't use a "standard" that isn't actually standard -- but messy datasets are all too common :-(

> So, users can deactivate any problematic variable from decoding, by passing a mapping as shown.

Sure, there are plenty of solutions, if you're thinking about what I call the "interactive data analysis" use case:

- try to open the dataset.
- if it fails, either fix the dataset, or give xarray special instructions to make it work.
   - in this particular case, I'm dropping the problematic variables -- not hard, but i have to write that code :-(

All good.

But if you are using xarray in a larger system, behind a web service, etc, it's really nice to be able to have things not fail when the failure is not a show stopper.

For instance, in this example the problematic variable is completely unnecessary. 

We already have `decode_times = False`, so you can keep going, having a way to decode all the times that it can, but not the ones it can't would be a nice addition.

> CFDatetimeCoder might actually be the right place to start, the suggested error-switch could just be added there (eg. on_error : {"raise", "warn", "ignore"} with raise as default). 

That step seems straightforward.

> The underlying implementation is NOT straightforward as we already have several warnings issued on the way (numpy.datetime vs cftime, calendar issues etc).

I can trace  it through and look at those, but as there are all those, maybe adding one more warning's not a big deal :-)

I'll try to make a start on a PR soon.

**Comment by kmuehlbauer** [comment]
@ChrisBarker-NOAA Thanks for the details. If you want to give it a go, I'm happy to review and test.

**Comment by ChrisBarker-NOAA** [comment]
Thanks! I started a WIP here: #11140

There are a few ways to go with passing the new flag on through open_dataset -- I'd like feedback on that, ideally before I try to implement it.

**Comment by kmuehlbauer** [comment]
I've pondered a bit on @dcherian's https://github.com/pydata/xarray/issues/2848#issuecomment-3845142355.

So, how about the following procedure:

- Acknowledge any proper CF time units string:  "**time_unit** **since** **reftime**" as to be decoded by `CFDatetimeCoder`.
- If either **time_unit** or **ref_time** does not resolve to a valid CF time units string, do not acknowledge it as valid CF time string and skip decoding for those variables.

The `on_error` mechanism could still be implemented as in #11140, but only for usage with `CFDatetimeCoder`.

This would mean:

- `decode_times=True` -> decode on conforming time units string, `skip` time decoding on non conforming units string
- `decode_times=False` -> `skip` time decoding entirely
- `decode_times=CFDatimeCoder(on_error=on_error)` -> `raise`, `warn`, `ignore`, `skip`
   - `raise` - try decode, raise error in case of any error (current behaviour)
   - `warn` - try decode, warn in case of any error, but return undecoded
   - `ignore` - try decode, return undecoded in case of any error
   - `skip` - skip decoding on non conforming string, return undecoded

In my opinion, and I might miss valid points of thought here, the `on_error` scheme would still add value to normal CF time decoding. In the same time non-conforming time units will just be skipped, actually helping immediately for occasions as @efiring and @ChrisBarker-NOAA mentioned. 

I think this also could go without deprecation, if we consider that we have an immediate solution for the OP's use case. And for those cases where this would have raised before, users can still apply their fixes to the time units string and decode again.

Finally, we would not force more meanings into `decode_times`. 

Thoughts?

**Comment by ChrisBarker-NOAA** [comment]
> If either time_unit or ref_time does not resolve to a valid CF time units string, do not acknowledge it as valid CF time string and skip decoding for those variables.

I'm not good with regex, but If I have this right, all it does is look for <something>since<something else>

and then assumes that <something> is a timedelta unit(e.g. days, seconds, etc) and <something else> is a datetime string (e.g. 2025-02-03 12:00:00).

And then, if either of those don't successfully parse as such, it raises an error.

@kmuehlbauer -- is your suggestion then that if either:
 - <something> isn't a valid timedelta unit
 - <something else> isn't a valid datetime string

then "skip" would be invoked? 

So what is the difference between "skip" and "ignore" ?

>  non-conforming time units will just be skipped

I guess I'm trying to understand what the definition is of "non-conforming time units" that's different that what we have now, which is think is:
  - Anything with a "since" in it that doesn't fully parse as a valid time unit'

**Comment by kmuehlbauer** [comment]
One of the first things we do in the decoding process is:

```python
time_unit, ref_date = _unpack_time_unit_and_ref_date(units)
```

which itself does two things:

1.
  - a. `time_unit, _ref_date = _unpack_netcdf_time_units(units)`
    - check for `since` inside `units` and split string accordingly - raise if fail -> should then skip decoding (we already skipped decoding if `since` is not present)
    - ensure year is padded - raise if fail (eg if no time string) -> should then skip decoding

  - b. `time_unit = _netcdf_to_numpy_timeunit(time_unit)`
    - check for proper CF time unit - raise if fail -> should then skip decoding
 
2.  `ref_date = pd.Timestamp(_ref_date)`  
  - raises if pd.Timestamp fails  
    - `1970-01-01 70:00:00` - erroneous time string - should then skip decoding
    - `10000-01-01 00:00:00` - out of bounds -> should be handled by cftime decoding

We could call these checks right in the beginning already in `CFDatetimeCoder.decode` to determine if we can skip, or if we should move into decoding. 

There is at least one pitfall in the whole scheme:

- using `pd.Timestamp` will raise for years >=10000 and <=-10000, even if it can be constructed from numpy.datetime64
  - `pd.Timestamp("10000-01-01 00:00:00")` -> raises ValueError/DateParseError
  - `pd.Timestamp(np.datetime64("10000-01-01 00:00:00", "s"))` -> works

`skip` in my interpretation will skip the whole processing (but only on non CF conforming units), whereas `ignore`, `warn` and `raise` will run the decoding algorithm and take the appropriate measures.

What are "Non CF conforming units"
----------------------------------

`<time_units> since <reference_time>`

- missing `since`
- erroneous/missing `time_units` (non of days, hours, minutes, seconds, milliseconds, microseconds or nanoseconds)
- erroneous `reference_time` (can't be parsed by `pd.Timestamp` and/or adapted functionality)

I'll try to come up with some code.

**Comment by kmuehlbauer** [comment]
The following code included in CFDatetimeCoder.decode will skip on invalid `time_units` and erroneous `reference_time`:

```python

      try:
          _time_unit, _ref_date = _unpack_netcdf_time_units(units)
          _time_unit = _netcdf_to_numpy_timeunit(_time_unit)
          _ref_date = pd.Timestamp(_ref_date)
      except KeyError as err:
          emit_user_level_warning(f"Skipping variable {name!r}, unable to decode time units: {err}")
          return variable
      except ValueError as err:
          msg = str(err).lower()
          # do not skip year out of range
          year_is_out_of_range = "is out of range" in msg and "year" in msg
          # skip invalid reference date, this cannot be repaired
          is_invalid_ref = "invalid reference date for time units" in msg
          if not year_is_out_of_range or is_invalid_ref:
              emit_user_level_warning(
                  f"Skipping variable {name!r}, unable to decode reference date: {err}")
              return variable

```

Note, that's just a proof of principle. We really would need to re-think using `pd.Timestamp` for ingestion of `reference_time`, as it will break on years >=10000 and <=-10000.

**Comment by ChrisBarker-NOAA** [comment]
> The following code included in CFDatetimeCoder.decode will skip on invalid time_units and erroneous reference_time:

What if both are bad? I think that was the case in my example. Should it skip or fail?

Anyway, I *think* this would change the current behavior -- which is fine with me, but would break code the relies on an Exception being raised for ANY problem with a time variable.

So I'm not sure that we can do this without another flag of some sort :-(

But as to what we want to do. I *think* the goal is to distinguish something that is not a CF time unt at all, from one that is intended to be a CF time unit, but is invalid in some way.

1) what is "intended" to be a cf time variable ?
 - that is -- something that the user intends to be a time variable, which may or may not be valid.
 - currently, I think we've defined that as "anything with a " since " in the units attribute, that is:

A unit string of the form: `"<unit> since <datetime>"``

I can't think of another definition we could use -- it seems you are now proposing defining it as:

A unit string of the form: `"<unit> since <datetime>"` *where `<unit>` is a valid time unit*.

But is the unit any more important / different than the datetime? If not, then we are kind of back to:

A cf time variable is anything that van be parsed as a valid CF datetime string -- which would be what I had called "ignore", and I think useful, but not backward compatible without a new flag.

So what to do?

One option is to use some other metric to identify what are intended to be CF time variables -- we could use standard_name, or whether it matches a time dimension, or something like that. However, that would probably reject some non-CF compliant variables that currently get decoded successfully :-( -- and it might well still identify ones with invalid time units anyway :-(

Personally, I'd be happy to make this more robust and skip more -- but I don't think there's a way to do that without either some new flag or breaking backward compatibility -- which is a no-no.

> We really would need to re-think using pd.Timestamp for ingestion of reference_time, as it will break on years >=10000 and <=-10000.

Is that relevant for any of the "standard" calendars? or does it only matter for climate science that should be using other calendars anyway (and thus cf_time)? -- though maybe people do do that.
 - I suppose we catch the exception, and try to use cf_time if ti raises?

**Comment by kmuehlbauer** [comment]
Thanks @ChrisBarker-NOAA for keeping the discussion going. Already excusing, this is going to be long.

> > The following code included in CFDatetimeCoder.decode will skip on invalid time_units and erroneous reference_time:
> 
> What if both are bad? I think that was the case in my example. Should it skip or fail?
> 
> Anyway, I _think_ this would change the current behavior -- which is fine with me, but would break code the relies on an Exception being raised for ANY problem with a time variable.
> 
> So I'm not sure that we can do this without another flag of some sort :-(

Difficult decision. We could move forward with just adding the `on_error` flag to `CFDatetimeCoder` and keep the current behaviour. We could additionally start deprecating the current default behaviour (for `decode_times=True`) to raise on malformed units and advertise usage of `CFDatimeCoder` `on_error`-flag. The new default behaviour after end of deprecation cycle would be to `ignore_noncf` induced errors and skip decoding of those variables. This would put some pressure downstream to adopt `CFDatetimeCoder` if they want to keep the behaviour. Which I'm totally fine with as it brings us more towards the original aim to split out coders completely.

> But as to what we want to do. I _think_ the goal is to distinguish something that is not a CF time unt at all, from one that is intended to be a CF time unit, but is invalid in some way.
> 
>     1. what is "intended" to be a cf time variable ?
> 
> 
>     * that is -- something that the user intends to be a time variable, which may or may not be valid.
> 
>     * currently, I think we've defined that as "anything with a " since " in the units attribute, that is:
> 
> 
> A unit string of the form: " since "
> 
> I can't think of another definition we could use -- it seems you are now proposing defining it as:
> 
> A unit string of the form: " since " _where is a valid time unit_.
> 
> But is the unit any more important / different than the datetime? If not, then we are kind of back to:
> 
> A cf time variable is anything that van be parsed as a valid CF datetime string -- which would be what I had called "ignore", and I think useful, but not backward compatible without a new flag.

OK, I think we have very well written down, what a valid CF time units string consists of in https://cfconventions.org/cf-conventions/cf-conventions.html#time-coordinate-units which boils down to `<time_units> since <reference_time>`.

For the `<time_units>` this is ~very~relatively easy to check (as we do), for the `<reference_time>` we rely on some prior checking before feeding to `pd.Timstamp`. We _could_ use these checks to at least sort out the most `broken`/`malformed` units strings, and assume that any of the surviving units are intended CF time units strings. Those found were the ones I would have put under `on_error="skip"`, better to name `on_error="ignore_noncf"`, which is different from @ChrisBarker-NOAA's `on_error="ignore"` (ignores any errors).

> Is that relevant for any of the "standard" calendars? or does it only matter for climate science that should be using other calendars anyway (and thus cf_time)? -- though maybe people do do that.
> 
>     * I suppose we catch the exception, and try to use cf_time if ti raises?

It's relevant for `proleptic_gregorian` which can extend for aeons in any direction already for `microsecond` resolution. Not sure how useful this is, but `np.datetime64` as well as `pd.Timestamp` are capable of this.

And yes, trying `cftime` is also one of the major downsides of the current logic, any error raised on the decoding path will run again using `cftime`. If `cftime` isn't installed it will suggest to do so. Without actually helping the user for malformed units as finally this raises in `cftime`. At least we always suggest to use `decode_times=False`.

So how to move forward? We need to answer the following questions (I'll include my answers and proposed strategy): 

- What should be default behaviour for `decode_times=True`?

  1. Decode everything which can be decoded
  2. Skip decoding for variables which we assume are not CF time variables/coordinates. That assumption is made by the checks prior feeding ``reference_time`` to `pd.Timestamp`. Need to document this properly.
  3. Still raise errors which come up in subsequent processing

- Should we deprecate current behaviour and if, how?

  1. We need to deprecate current behaviour, if we want to follow the road of splitting out coders and also to have a more fine-grained error handling.
  2. To deprecate, emit warnings before raising errors which would be raised by the most `broken`/`malformed` units strings (see above for explanation). Something along the lines:
    `"This will no longer be raised in future versions. To silence this warning use decode_times=CFDatetimeCoder(on_error="raise") to keep current behaviour or decode_times=CFDatetimeCoder(on_error="ignore_noncf") to adopt to new default."`
  3. design `on_error`:
    ```python
    on_error : {'raise', 'warn', 'ignore', 'ignore_noncf'}, defaults to 'raise'
        Specifies how to handle errors during coding:

        - 'raise' : Raise all errors.
        - 'warn' : Emit a warning for any error and return original.
        - 'ignore' : Silently ignore all errors and return original.
        - 'ignore_noncf' : Silently ignore **non-CF-compliant units** and return original. 
          Other errors are still raised.
    ```
  4. To finalize deprecation, set `on_error` default to "ignore_noncf".

Although this will not immediately help for @ChrisBarker-NOAA or @efiring use cases with a simple `decode_times="ignore", but `decode_times=CFDatetimeCoder(on_error="ignore")` is not that ugly, too?

Setting this once in downstream code does also look usable: 

```python
import xarray as xr
coder = xr.coders.CFDatetimeCoder(on_error="ignore")
ds0 = xr.open_dataset(filename0, decode_times=coder)
ds1 = xr.open_dataset(filename1, decode_times=coder)
```

To get a rough number on deprecation time frame, we could finalize deprecation of `use_cftime` now and see how this goes. Maybe we only need half a year to inform users. Would this be something people could agree on?

**Comment by ChrisBarker-NOAA** [comment]
> And yes, trying cftime is also one of the major downsides of the current logic, any error raised on the decoding path will run again using cftime. If cftime isn't installed it will suggest to do so. 

I note here -- I"m pretty sure that I've gotten a error with a suggestion to install cftime even though cftime was in fact already installed.

I'll try to see if I can find where that happened and clean that up.

**Comment by ChrisBarker-NOAA** [comment]
> Although this will not immediately help for @ChrisBarker-NOAA or @efiring use cases with a simple decode_times="ignore", but decode_times=CFDatetimeCoder(on_error="ignore")` is not that ugly, too?

Not too bad -- and better than what we have now -- at least there's some fairly easy way to get what we need :-)

So as a next step -- that would be great.

> And yes, trying cftime is also one of the major downsides of the current logic, any error raised on the decoding path will run again using cftime. 

That does complicate it, but I don't know that it makes the API harder to do -- either way, I'm not sure there's a way to clearly identify unit strings that are *supposed to be* valid, but have an error, vs one that looks a bit like one, but isn't intended to one. So we need a simple metric for "looks like a time unit" (e.g. the <something>since<something else> strucured used now) -- and only ave to decide what to do -- and an easy way to specify the choice.

> We could move forward with just adding the on_error flag to CFDatetimeCoder

Good first step.

> We could additionally start deprecating the current default behaviour (for decode_times=True) to raise on malformed units and advertise usage of CFDatimeCoder on_error-flag. 

Or follow my suggestion for open_dataset -- keep the current behavior, and add addition flags:

decode_times : {True, False, 'never', 'raise', 'warn', 'ignore', 'ignore_noncf'}, defaults to 'raise'

True and False are there for backward compatibility:

True == 'raise' 
False == 'never'

Deprecate True and False eventually (or never, it's not SO bad to have duplicates in the API)

I think this is a much better API for users than having to specify a Decoder.

But if you really son't like it, then at least we'll have the option once the flags have been added to the decoder.

**Comment by kmuehlbauer** [comment]
> But if you really son't like it, then at least we'll have the option once the flags have been added to the decoder.

It's not a matter of liking, but it does not make sense to me to add things in one place, when we are working to split things out in other places. But that's just my POV and I'm open to other views here. For #11140 I think it would be good to move forward with the implementation of `on_error` in `CFDateTimeCoder`.

**Comment by dcherian** [comment]
I am v. confused. https://github.com/pydata/xarray/issues/2848#issuecomment-3844695073 has bad units: `"msec since 00:00:00"` . Can we just skip decoding here and raise a warning (as we do for other cases)?

**Comment by kmuehlbauer** [comment]
That's what I suggested, that we skip encoding if our checks for time_unit and refrerence_time fail.

The issue is, that we might have to deprecate the current behaviour of raising as downstream is relying on this.

 @dcherian  If we assume this as bugs, we could just skip and warn as you suggested.

But for other issues in the decoding the on_error would still be useful.

**Comment by ChrisBarker-NOAA** [comment]
@dcherian: yes, that's exactly what we want. The problem is backward compatibility.

Also -- different users will want different behavior.

Anyway, I *think* think there is consensus (among the three? of us) that there should be three options:

"skip", "raise", "warn" for CF decoding. And I've added that in this PR.

The trick is how to express that in `open_dataset` with a clean API that's backward compatible ...

> skip encoding if our checks for time_unit and refrerence_time fail.
...
> But for other issues in the decoding the on_error would still be useful.

I don't know that there's any way to know what "kind" of error it is.
First step:
- Does this look like it's intended to be a CF-style time variable ?
   - looking for "<something1> since <something2>" is currently the test for that, and I can't think of another one.
Second step:
- Try to decode:
   - <something1> as a time unit
   - <something2> as a datetime

If either of these fails, then it's an invalid CF time string.

I don't think we should handle this differently depending on which of those failed (except a smarter error message).

In short:

It's either intended to be a time variable, or it's not

and

If so, it's either decodable or it's not

*maybe* someone smarter than me can come up with a way to decide that it's not intended to be a CF time variable even when it has the structure <something1> since <something2>, but I can't think of one.

**Comment by ChrisBarker-NOAA** [comment]
Thinking a bit more, here's my proposal for open_dataset:

```
decode_times: str, CFDatetimeCoder or dict-like, optional

str must be one of:  {"never", "raise", "warn", "ignore"}

Default: "raise"

If the flag is "never" no time decoding will be performed.

If the string flag is not "never", variables that appear to be encoded in the standard NetCDF datetime format with be decoded into datetime objects. If the decoding fails for a variable, either raise an exception, raise a warning, or ignore decoding of that variable.

Otherwise, use [coders.CFDatetimeCoder](https://docs.xarray.dev/en/stable/generated/xarray.coders.CFDatetimeCoder.html#xarray.coders.CFDatetimeCoder) or leave them encoded as numbers.

Pass a mapping, e.g. {"my_variable": False}, to toggle this feature per-variable individually. This keyword may not be supported by all the backends.
```

`True` and `False` would be deprecated, but still work for a LONG time (forever?):

'True == "raise"`

`False == "never"`

[not sure "never" is the best way to spell it, open to ideas there]

Whether to document the `True` and `False` as deprecated (in the docstring)  for a while is up in the air...

We could also not deprecate `True` and `False` -- but why cement in an ugly interface forever?


NOTE: if PR #11140 is done, then we could get this functionality by passing a  `coders.CFDatetimeCoder` in. However, I think the goal of the `open_dataset()` API is to be interactive-use friendly, and a string flag is a lot friendlier for that use that requiring another import and understanding (at least a little) how decoders work ..

**Comment by kmuehlbauer** [comment]
@ChrisBarker-NOAA I'm with you that the user facing API of `open_dataset` should be fairly easy. 

To add a bit more here, the long term goal is to remove any cf-related keyword arguments from the `open_` API but keep a single `decode`-kwarg (or whatever is decided on). This was suggested by @shoyer in #4490. The implementation of `CFDatetimeCoder` and deprecation of `use_cftime`-kwarg was only a first step.

So everything which is added/changed to the API should be made with the above aim in mind.

> * looking for " since " is currently the test for that, and I can't think of another one.

That is what we expect: `<time_unit> since <reference_datetime>`

These are clearly no CF time units:
```
"meter since start of measurement"
"foo since bar"
"bananas since yesterday"
```

These are also not CF compliant (see #4927):
```
hours since analysis (@efiring)
seconds since reference_time (@kmuehlbauer)
seconds since measurement_start
```

This is also not CF compliant:
```
msec since 00:00:00 (@ChrisBarker-NOAA)
```

For `<time_unit>` we can check against an explicit whitelist of CF-valid time units (we do this already). For `<reference_datetime>` we check currently check if we can extract four digit year or at least pad to four digits. These tests could be used to differentiate if a `units`-attribute is intended a CF-time units.

So my suggestion still stands (as written in https://github.com/pydata/xarray/issues/2848#issuecomment-3884259525), deprecate current behaviour and skip decoding in the above cases. A discussion of that topic in the xarray dev meeting could help to find a way forward. I'll put it on the agenda.

**Comment by ChrisBarker-NOAA** [comment]
> So everything which is added/changed to the API should be made with the above aim in mind.

Got it -- I don't *think* that changes much for this particular example, it would be the `datetime` parameter of the `CFCoders` class (or dict) rather than the `decode_times` parameter of `open_dataset()` but otherwise the same.

Though the good news is that if you go that route, maybe backward compatibility isn't necessary for that parameter.

> That is what we expect: <time_unit> since <reference_datetime>

Sure, but if <time_unit> or <reference_datatime> are not valid, there is no way to distinguish between something that is intended to be cf time unit that has an error, and one that is intended to be something else.

(humans can do it -- clearly "meters since home" is not intended to be a time unit, but computers? not so much, not without a lot of logic.

And I *think* that's what the current code is trying to do. And it's a good idea -- if someone does have a typo in their units attribute that makes it undecodable, it's kind of nice to get an error or warning, rather than it simply being skipped.

I don't have an example of this on hand, but a similar one -- I recently dealt with a dataset that had:

time:calendar = "proleptic gregorian"  -- note the space, rather than the required underscore. That yielded a very understandable error -- but if it has simply been skipped, then it would have been very mysterious!

Here's a made up example:

`"miliseconds since 2025-12-12T00:00:00"`

(note the missing "l")

If I try to decode that with cftime, I get:

```
ValueError: In general, units must be one of 'microseconds', 'milliseconds', 'seconds', 'minutes', 'hours', or 'days' (or select abbreviated versions of these).  For the '360_day' calendar, 'months' can also be used, or for the 'noleap' calendar 'common_years' can also be used. Got 'miliseconds' instead, which are not recognized.
```

Which is great -- that's what we want the user to see -- in an error, or a warning.

So we don't want to suppress those (skip them) unless the user has explicitly asked that they be skipped.

> So my suggestion still stands (as written in https://github.com/pydata/xarray/issues/2848#issuecomment-3884259525), deprecate current behaviour and skip decoding in the above cases. A discussion of that topic in the xarray dev meeting could help to find a way forward.

So I'm suggesting we don't skip decoding in those cases unless the user explicitly asks for that.

NOTE: I get the same error with `"meters since my_house"`

And with `"seconds since my_house"`: I get: 

`ValueError: Unable to parse date string 'my_house'`

also a reasonable error.

NOTE: 
> explicit whitelist of CF-valid time units (we do this already)

CF is a bit ungainly in this regard: "The values of the units attributes are character strings that are recognized by UNIDATA’s UDUNITS package" -- and UDUNITS kinda follows Postel's Law: "...be liberal in what you accept from others".

That is, there is no whitelist that I know of anywhere  and if there were, it would be big. I have yet to find a clear documentation of the rules (e.g. plural vs singular, etc). The only way I know to do it for sure is to use UDUNITS itself:

```
You have: 5 s
You want: hours
    5 s = 0.00138889 hours
    x/hours = 0.000277778*(x/s)
You have: 5 second
You want: hour
    5 second = 0.00138889 hour
    x/hour = 0.000277778*(x/second)
You have: 5 seconds   
You want: hour
    5 seconds = 0.00138889 hour
    x/hour = 0.000277778*(x/seconds)
You have: 5 kiloseconds
You want: hour
    5 kiloseconds = 1.38889 hour
    x/hour = 0.277778*(x/kiloseconds)
```
I don't know that anyone ever uses "kiloseconds" but it is CF compliant :-(

It's reasonable enough for xarray to have a large, but constrained whitelist, but it wouldn't quite be CF :-(

I note that cftime doesn't require UDUNITS -- it's using a whitelist.
```
In [5]: cftime.num2date(100000, "kiloseconds since 2025-12-12T00:00:00")
---------------------------------------------------------------------------
ValueError: In general, units must be one of 'microseconds', 'milliseconds', 'seconds', 'minutes', 'hours', or 'days' (or select abbreviated versions of these).  For the '360_day' calendar, 'months' can also be used, or for the 'noleap' calendar 'common_years' can also be used. Got 'kiloseconds' instead, which are not recognized.

```

xarray following cftime is perfectly reasonable. Though *maybe* we could add a note to the error message that it doesn't fully support ALL legal units.

**Comment by kmuehlbauer** [comment]
Hi @ChrisBarker-NOAA 👋, we've discussed that shortly in todays dev-meeting.

- plain "since" check is OK
- adding meaning to `open_dataset` `decode_times`-kwarg is OK, if backwards-compatible

As I still see added value in a kind of non-cf check (explained extensivly in my above comments), this could go alongside your {"raise", "warn", "ignore"} as eg. "ignore_noncf". This doesn't necessarily have to be within your PR. 

Thanks for pushing this!

**Comment by ChrisBarker-NOAA** [comment]
Thanks! I'll try to get that all done soon.

> eg. "ignore_noncf". 

I'm confused as to how that would be different that "ignore"? Or is this specifically about stuff that cftime supports that is not otherwise supported (e.g. non-standard calendars) ?

> This doesn't necessarily have to be within your PR.

OK then -- I'll leave that out for now :-)

**Comment by ChrisBarker-NOAA** [comment]
I've got a complete draft of this in #11140  -- finally!
